README.md
# Community Platform

Run a simplified JWA platform on your local machine:

1. Learn about the platform and see what it can do
1. Integrate your applications and games with the platform
1. Discover new usages and share them with the community

## Stack

The community platform starts the core JWA Platform services:

1. The NATS.io message bus
1. The REST ingress service (airlock)
1. A mock authorization service for airlock
1. A Tezos sandbox
1. Observability tools

It then starts additional services:

1. JWA Item Service
1. Or run your own services on top of it

## Getting started

To get started, all you need is a recent version of Docker, including Docker compose.

```
npm install -g @jwalab/community-platform

jwalab --help
```

Start the platform:

```
jwalab start
```

## Tutorial

An online tutorial is available here: https://jwa-lab.github.io/platform-docs/

## Options

### Changing ports

It's possible that you have other applications running on ports that the Community Platform may want to use.

Here's the list of ports that will be mapped to your local machine:

| Service       | Default Port | Description                                                                          | Override env variable    |
| ------------- | ------------ | ------------------------------------------------------------------------------------ | ------------------------ |
| Tezos Sandbox | 20000        | The local Tezos Node. You may make RPC calls to this URL                             | JWALAB_TZBOX_PORT        |
| NATS          | 4222         | NATS is the message bus. Port 4222 is the main port for pub/sub                      | JWALAB_NATS_PORT         |
| Auth Service  | 8999         | The Mock Auth Service that returns Bearer Tokens to authenticate with Airlock        | JWALAB_AUTH_SERVICE_PORT |
| Airlock       | 8000         | The Platform's main port of entry, you will make all of the APIs calls via this port | JWALAB_AIRLOCK_PORT      |
| TzStats       | 8001         | A useful tool for collecting stats on the Tezos Sandbox                              | JWALAB_TZSTATS_PORT      |
| TzIndex       | 8002         | A useful tool for exploring the Tezos Sandbox                                        | JWALAB_TZINDEX_PORT      |
| PostgreSQL    | 5432         | The platform's DB for storing items, useful for debugging                            | JWALAB_POSTGRES_PORT     |
| Swagger UI    | 3000         | A Swagger UI for navigating the platform's documentation                             | JWALAB_SWAGGER_PORT      |

If you'd like to remap any of those ports, simply set the port value to the corresponding environment variable override.

For instance, setting Airlock to a new port:

```
JWALAB_AIRLOCK_PORT=9000 jwalab start
```

### .env

The platform currently doesn't support reading environment variables from a .env file, however, feel free to create one and store it where you want. You can load it an run the platform using this syntax:

```sh
export $(cat .env) && jwalab start
```

### Currently used ports

There's a useful command to list all the ports being used by the platform, along with the environment variable to remap the port.

```sh
$ jwalab list-services

┌─────────┬─────────────────┬───────────────┬────────────────────────────┐
│ (index) │  Service Name   │  Mapped Port  │    ENV VAR for override    │
├─────────┼─────────────────┼───────────────┼────────────────────────────┤
│    0    │    'Airlock'    │ 'Not running' │   'JWALAB_AIRLOCK_PORT'    │
│    1    │ 'Auth Service'  │     8999      │ 'JWALAB_AUTH_SERVICE_PORT' │
│    2    │     'NATS'      │     4222      │     'JWALAB_NATS_PORT'     │
│    3    │  'PostgreSQL'   │     5432      │   'JWALAB_POSTGRES_PORT'   │
│    4    │  'Swagger UI'   │     8080      │   'JWALAB_SWAGGER_PORT'    │
│    5    │ 'Tezos Sandbox' │     20000     │    'JWALAB_TZBOX_PORT'     │
│    6    │    'TzIndex'    │     8002      │   'JWALAB_TZINDEX_PORT'    │
│    7    │    'TzStats'    │     8001      │   'JWALAB_TZSTATS_PORT'    │
└─────────┴─────────────────┴───────────────┴────────────────────────────┘
```

## Commands

```
JWA_LAB - v0.1
Run a JWA Community platform (jwalab) on your local machine!


Options:
  -V, --version   output the version number
  -h, --help      display help for command

Commands:
  start           start the jwalab environment
  stop            stop the jwalab environment
  kill            kill the jwalab environment
  pull            download the latest versions of the platform's services
  tezos-client    run a command in the tezos-client
  granabox        run a command in granabox
  list-services   list all running services and the port they expose
  logs            view platform logs
  help [command]  display help for command
```


CHANGELOG.md
# Changelog

## 0.1.12

-   Update to `item-service:0.0.11` and `tezos-gateway:0.0.4` which replaces `tezos-work-queue`.
    The upgrade brings significant performance improvement during tokenization by batching operations together.

## 0.1.11

-   Update to `authorization-service:0.0.5` which allow studio_id and user_id update for development tokens.

## 0.1.10

-   Upgrade item-service and airlock to most recent path versions

## 0.1.9

-   Update to `item-service:0.0.5` which adds the freeze item endpoint.
-   Update to `tezos-work-queue:0.0.2`

## 0.1.8

-   Update to `item-service:0.0.4`, automates `available_quantity`.

## 0.1.7

-   Fix `tezos-client` and `granabox` commands

## 0.1.6

-   Fix issue with `list-services`

## 0.1.5

-   Add a `logs` command to view all logs across all services

## 0.1.4

-   Rename list-ports to list-services and fix output when the platform is not running

## 0.1.3

-   Add option to remap ports used by the community platform
-   Add command to list which ports are being used by the platform

## 0.1.1

-   Upgrade item-service to 0.0.2 which includes improved logic for recomposing item instances

## 0.1.0

-   [BREAKING] Removing `item-store` and `tokenization-service` and replace with newer `item-service`
    and `tezos-work-queue` combination which provide a simpler, integrated API and an asynchronous work queue.
    See documentation for more details: https://jwa-lab.github.io/platform-docs/

## 0.0.22

-   Upgrade authorization-service to 0.0.4

## 0.0.21

-   Upgrade authorization-service to 0.0.3

## 0.0.20

-   Upgrade airlock to 0.0.8

## 0.0.19

-   Upgrade to Granada protocol

## 0.0.18

-   Upgrade airlock to 0.0.7
-   Upgrade authorization-service to 0.0.2
-   Upgrade item-store to 0.0.8
-   Upgrade tokenization-service to 0.0.10

## 0.0.17

-   Set Tezos block_time to 2s to speed up development and testing

## 0.0.16

-   Upgrade protocol to Florence
-   Add mock authorization service and require Auth token to access Airlock API


index.js
#!/usr/bin/env node

const { Command } = require("commander");
const clear = require("clear");
const compose = require("docker-compose");
const path = require("path");
const chalk = require("chalk");
const { spawn } = require("child_process");

const servicesConfig = require("./servicesConfig.js");
const package = require("./package.json");

clear();

ensurePorts(servicesConfig);

const { BOX_NAME = "granabox", JWALAB_TZBOX_PORT } = process.env;

const program = new Command();

program
    .description(
        `${chalk.grey("JWA")}${chalk.bold("_LAB")} - v${package.version}
Run a JWA Community platform (jwalab) on your local machine!
`
    )
    .version(package.version)
    .name("jwalab a.k.a JWA Community platform")
    .usage("[global options] command");

program
    .command("start")
    .description("start the jwalab environment")
    .action(async () => {
        try {
            await compose.upAll({ cwd: path.join(__dirname), log: true });
        } catch (err) {
            console.error("something went wrong:", err);
        }
    });

program
    .command("stop")
    .description("stop the jwalab environment")
    .action(() => {
        compose
            .down({
                cwd: path.join(__dirname),
                log: true,
                commandOptions: ["--remove-orphans"]
            })
            .then(
                () => {
                    console.log("jwalab stopped");
                },
                (err) => {
                    console.log("something went wrong:", err.message);
                }
            );
    });

program
    .command("kill")
    .description("kill the jwalab environment")
    .action(() => {
        compose
            .kill({
                cwd: path.join(__dirname),
                log: true,
                commandOptions: ["--remove-orphans"]
            })
            .then(
                () => {
                    console.log("jwalab killed");
                },
                (err) => {
                    console.log("something went wrong:", err.message);
                }
            );
    });

program
    .command("pull")
    .description(`download the latest versions of the platform's services`)
    .action(() => {
        compose
            .pullAll({
                cwd: path.join(__dirname),
                log: true
            })
            .then(
                () => {
                    console.log("jwalab updated");
                },
                (err) => {
                    console.log("something went wrong:", err.message);
                }
            );
    });

program
    .command("tezos-client")
    .description("run a command in the tezos-client")
    .action(() => tezosClient(program.args.slice(1).join(" ")))
    .addHelpText(
        "after",
        `
    Examples:
      $ jwalab tezos-client -- --help
      $ jwalab tezos-client -- list known contracts
`
    );

program
    .command(BOX_NAME)
    .description(`run a command in ${BOX_NAME}`)
    .action(() => runBox(BOX_NAME, program.args.slice(1).join(" ")))
    .addHelpText(
        "after",
        `
    Examples:
      $ jwalab ${BOX_NAME} -- --help
      $ jwalab ${BOX_NAME} -- info
`
    );

program
    .command("list-services")
    .description("list all running services and the port they expose")
    .action(async () => {
        const servicePorts = [];

        await Promise.allSettled(
            Object.entries(servicesConfig).map(async ([name, { port }]) => {
                const servicePort = {
                    "Service Name": servicesConfig[name].name,
                    "Mapped Port": "Not running",
                    "ENV VAR for override": servicesConfig[name].overrideVar
                };

                try {
                    const { out } = await compose.port(name, port, {
                        cwd: path.join(__dirname)
                    });

                    servicePort["Mapped Port"] = Number(out.split(":")[1]);
                } catch (err) {}

                servicePorts.push(servicePort);
            })
        );

        console.table(
            servicePorts.sort((serviceA, serviceB) =>
                serviceA["Service Name"] > serviceB["Service Name"] ? 1 : -1
            )
        );
    });

program
    .command("logs")
    .description("view platform logs")
    .action(async () => {
        const log = spawn("docker", ["compose", "logs", "-f"], {
            cwd: path.join(__dirname),
            maxBuffer: 1_000_000
        });

        log.stdout.on("data", (data) => console.log(String(data)));
        log.stdout.on("error", (err) => console.error("error: ", err));
    });

program.parse(process.argv);

function tezosClient(command) {
    return compose.exec(
        "tzbox",
        `tezos-client -E http://localhost:${JWALAB_TZBOX_PORT} ${command}`,
        {
            cwd: path.join(__dirname),
            log: true
        }
    );
}

function runBox(boxName, command) {
    return compose.exec("tzbox", `${boxName} ${command}`, {
        cwd: path.join(__dirname),
        log: true
    });
}

function ensurePorts(services) {
    Object.values(services).forEach(({ overrideVar, port }) => {
        if (!process.env[overrideVar]) {
            process.env[overrideVar] = port;
        }
    });
}


servicesConfig.js
module.exports = {
    tzbox: {
        name: "Tezos Sandbox",
        overrideVar: "JWALAB_TZBOX_PORT",
        port: 20000
    },
    nats: {
        name: "NATS",
        overrideVar: "JWALAB_NATS_PORT",
        port: 4222
    },
    "authorization-service": {
        name: "Auth Service",
        overrideVar: "JWALAB_AUTH_SERVICE_PORT",
        port: 8999
    },
    airlock: {
        name: "Airlock",
        overrideVar: "JWALAB_AIRLOCK_PORT",
        port: 8000
    },
    tzindex: {
        name: "TzIndex",
        overrideVar: "JWALAB_TZINDEX_PORT",
        port: 8002
    },
    tzstats: {
        name: "TzStats",
        overrideVar: "JWALAB_TZSTATS_PORT",
        port: 8001
    },
    postgres: {
        name: "PostgreSQL",
        overrideVar: "JWALAB_POSTGRES_PORT",
        port: 5432
    },
    "swagger-ui": {
        name: "Swagger UI",
        overrideVar: "JWALAB_SWAGGER_PORT",
        port: 8080
    }
};
